### ISSUE TITLE
Finish Event fired twice per request

### ISSUE THREAD

**Issue by mrmonat** [inertiajs/inertia#305]
There seems to be a bug, where the ```finish``` event gets called twice for a single Inertia get request.

I have the following Vue2 form watcher:
```javascript
watch: {
        form: {
            handler: debounce(function () {
                console.log('form-change')
                this.loading = true
                this.$inertia.get('/foo', {
                    filter: this.form.filter, // query params
                }, {
                    preserveScroll: true,
                    preserveState: true,
                    replace: true,
                    onFinish: () => {
                        console.log('finish')
                        this.loading = false
                    },
                    onSuccess: () => {
                        console.log('success')
                    }
                })
            }, 250),
            deep: true,
        }
    },
```

The first time the ```form```-value changes, the output in the console looks like the following:
```javascript
form-change
success
finish
```

Any additional change of the ```form```-value leads to the following output:
```javascript
// second change of "form"
form-change
finish // instant after form has changed
success // after inertia got a response from backend
finish // after inertia got a response from backend

// third change of "form"
form-change
finish // instant after form has changed
success // after inertia got a response from backend
finish // after inertia got a response from backend
```

So as the console output shows the ```onSuccess``` callback seems to get called only once per request (as expected), but the ```onFinish``` callback gets called twice.

**Comment by reinink** [comment]
What version of Inertia are you using?

**Comment by reinink** [comment]
Never mind, I am able to reproduce in the latest version. Let me try figure out what's going on here. 👍

**Comment by ajnsn** [comment]
@reinink Hey Jonathan. There might be still an issue with this `onFinish` event triggered twice in the most current Inertia version (v0.10.0). 

This seems an edge case and it seems to occur when two visits "overlap" each other. At least when using the per-visit callback.

If you want to have a look, I created a demo repo [over here](https://github.com/ajnsn/inertia-onfinish). There's only one route inside the [web.php](https://github.com/ajnsn/inertia-onfinish/blob/main/routes/web.php). It will serve the component with`triggerReload` set to `false` on the first visit. The second visit will make this prop set to `true`.

The component simply provides a link to visit this route again and has  a watcher on `triggerReload` which immediately fires another visit to the same route. 

There will be two lines with `onFinish` logged to the console.

As this is a per-visit callback, it should stick to this last visit and only log the `onFinish` event once, right?
I can mitigate this issue by wrapping the third visit in a `setTimeout`:

```js
reload() {
    setTimeout(() => {
        this.$inertia.visit('/', {
            preserveState: true,
            onBefore: () => {
                console.log('onBefore');
            },
            onFinish: () => {
                console.log('onFinish');
            },
        });
    });
}
```

This will only log one `onFinish` to the console.

Hope this helps!

**Comment by reinink** [comment]
Thanks @ajnsn. Will dig into this when I have some time. 👍

**Comment by robsontenorio** [comment]
@ajnsn have you solve it? I am on `0.10.1` and that still is an issue .

**Comment by ajnsn** [comment]
Hey @robsontenorio 
Still an issue and `setTimeout` works pretty well without a noticeable delay. Before that I used `async/await` and avoided the `onFinish` callback -  but then Inertia stopped returning the Axios promise.

Simple things like `loading = true` in `onStart` and `loading = false` in `onFinish` still work not as expected.

**Comment by Geelik** [comment]
Same issue with inertia 1.0.0 and vue 3. I'm doing a partial reload inside `onMounted `callback.

**Comment by reinink** [comment]
Hey! Thanks so much for your interest in Inertia.js and for sharing this issue/suggestion.

In an attempt to get on top of the issues and pull requests on this project I am going through all the older issues and PRs and closing them, as there's a decent chance that they have since been resolved or are simply not relevant any longer. My hope is that with a "clean slate" me and the other project maintainers will be able to better keep on top of issues and PRs moving forward.

Of course there's a chance that this issue is still relevant, and if that's the case feel free to simply submit a new issue. The only thing I ask is that you please include a super minimal reproduction of the issue as a Git repo. This makes it much easier for us to reproduce things on our end and ultimately fix it.

Really not trying to be dismissive here, I just need to find a way to get this project back into a state that I am able to maintain it. Hope that makes sense! ❤️

**Comment by CWAscend** [comment]
Hi @reinink, firstly thank you for making my life and our development team infinitely better with all the inertia goodness you've been delivering.

Just wanting to flag this issue to you again, we ran into the same issue yesterday.

We're on inertia v1.2, vue v3.4.27.

Our scenario is we're loading a page, then immediately partial reloading a lazy prop for snappiness, because the data in the lazy prop is being pulled through from an external API.

For some reason our onFinish seems to be triggering twice, messing with our loading state, which is immediately being set to false regardless of the actual partial reload still pending finish.

Like people have described in this issue,  setTimeout of 1ms is fixing it for us, which just gives me the shivers 😆 

If you need any more info from me I'm happy to give them.
